rejex.test.js

```diff
@@ -0,0 +1,170 @@
+});
+
+test('alternation short', () => {
+  const expr =
+    new AlternationExpr(  // [he]
+      new CharMatchExpr("h"),
+      new CharMatchExpr("e"))
+  expect(regexMatch("", expr)).toBe(false);
+  expect(regexMatch("h", expr)).toBe(true);
+  expect(regexMatch("e", expr)).toBe(true);
+  expect(regexMatch("he", expr)).toBe(false);
+});
+
+test('alternation long', () => {
+  // [hello]
+  const expr =
+    new AlternationExpr(  // (hell)o
+      new AlternationExpr(  // (he)(ll)
+        new AlternationExpr(  // he
+          new CharMatchExpr("h"),
+          new CharMatchExpr("e")),
+        new AlternationExpr(
+          new CharMatchExpr("l"),
+          new CharMatchExpr("l"))),
+      new CharMatchExpr("o"))
+  expect(regexMatch("h", expr)).toBe(true);
+  expect(regexMatch("e", expr)).toBe(true);
+  expect(regexMatch("l", expr)).toBe(true);
+  expect(regexMatch("o", expr)).toBe(true);
+  expect(regexMatch("x", expr)).toBe(false);
+  expect(regexMatch("hello", expr)).toBe(false);
+  expect(regexMatch("", expr)).toBe(false);
+});
+
+test('repetition short', () => {
+  const expr =
+    new RepetitionExpr(  // h*
+      new CharMatchExpr("h"))
+  expect(regexMatch("", expr)).toBe(true);
+  expect(regexMatch("h", expr)).toBe(true);
+  expect(regexMatch("hh", expr)).toBe(true);
+  expect(regexMatch("hhh", expr)).toBe(true);
+  expect(regexMatch("hhhh", expr)).toBe(true);
+});
+
+test('repetition wrong char', () => {
+  // h*
+  const expr = new RepetitionExpr(new CharMatchExpr("h"))
+  expect(regexMatch("g", expr)).toBe(false);
+});
+
+test('complex repetition', () => {
+  const expr =
+    new RepetitionExpr(  // [he]*
+      new AlternationExpr(  // [he]
+        new CharMatchExpr("h"),
+        new CharMatchExpr("e")))
+  expect(regexMatch("", expr)).toBe(true);
+  expect(regexMatch("h", expr)).toBe(true);
+  expect(regexMatch("hhhhh", expr)).toBe(true);
+  expect(regexMatch("e", expr)).toBe(true);
+  expect(regexMatch("eeeee", expr)).toBe(true);
+  expect(regexMatch("heheh", expr)).toBe(true);
+  expect(regexMatch("hefeh", expr)).toBe(false);
+});
+
+test('alternation is greedy', () => {
+  const expr =
+    new AlternationExpr(  // h*|el
+      new RepetitionExpr(  // h*
+        new CharMatchExpr("h")),
+      new ConcatExpr(  // el
+        new CharMatchExpr("e"),
+        new CharMatchExpr("l")))
+  expect(regexMatch("", expr)).toBe(true);
+  expect(regexMatch("hh", expr)).toBe(true);
+  expect(regexMatch("hhh", expr)).toBe(true);
+  expect(regexMatch("e", expr)).toBe(false);
+  expect(regexMatch("el", expr)).toBe(true);
+  expect(regexMatch("ell", expr)).toBe(false);
+  expect(regexMatch("ello", expr)).toBe(false);
+});
+
+test('h[e3]l*o|wo[r4]ld', () => {
+  const expr =
+    new AlternationExpr(  // h[e3]l*o|wo[r4]ld
+      new ConcatExpr(  // (h[e3]l*)o
+        new ConcatExpr(  // (h[e3])l*
+          new ConcatExpr(  // h[e3]
+            new CharMatchExpr("h"),
+            new AlternationExpr(  // [e3]
```

Comment:
Point out to everyone how the top level AlternationExpr is using a pipe | versus not using a pipe in square brackets [] at the CharMatchExpr level. And perhaps either add another test case or tell everyone an example where there may be more than one pipe in an expression and how that might look and be understood?

Reply:
I'll mention that you can use them multiple times... I'm not sure we'll even get to this last example though.